lte/cloud/go/services/subscriberdb/servicers/subscriberdb_servicer.go

```diff
@@ -89,13 +94,77 @@ func (s *subscriberdbServicer) ListSubscribers(ctx context.Context, req *lte_pro
 		subProto.NetworkId = &protos.NetworkID{Id: networkID}
 		subProtos = append(subProtos, subProto)
 	}
+
+	// Short-circuit subscriber data transmission by comparing cloud & gateway digests.
+	// If digest generation fails, the error is swallowed to not affect the main functionality.
+	//
+	// Note: currently the logic only applies to a request for the first page to avoid concurrency issues;
+	// more to be implemented with gateway subscriber cache that will canonize cloud digests.
+	noUpdates, digest := false, req.PreviousDigest
+	if req.PageToken == "" {
+		digest, err = getDigest(gateway, apnsByName, apnResourcesByAPN)
+		if err != nil {
+			glog.Errorf("Generating digest for subscribers in network of gateway %s failed", gatewayID)
+		} else {
+			noUpdates = req.PreviousDigest == digest
+			if noUpdates {
+				subProtos = []*lte_protos.SubscriberData{}
+			}
+		}
+	}
+
 	listRes := &lte_protos.ListSubscribersResponse{
 		Subscribers:   subProtos,
 		NextPageToken: nextToken,
+		Digest:        digest,
+		NoUpdates:     noUpdates,
 	}
 	return listRes, nil
 }
 
+// getDigest loads all subscribers registered on the current network, and returns
+// a deterministic subscriber flat digest.
+func getDigest(
+	gateway *protos.Identity_Gateway,
+	apnsByName map[string]*lte_models.ApnConfiguration,
+	apnResourcesByAPN lte_models.ApnResources,
+) (string, error) {
+	networkID := gateway.GetNetworkId()
+	lc := configurator.EntityLoadCriteria{
+		PageSize:           0,
+		PageToken:          "",
+		LoadConfig:         true,
+		LoadAssocsToThis:   true,
+		LoadAssocsFromThis: true,
+	}
+	subEnts, _, err := configurator.LoadAllEntitiesOfType(
+		networkID, lte.SubscriberEntityType, lc, serdes.Entity,
+	)
+	if err != nil {
+		return "", errors.Wrapf(err, "load all subscribers in network of gateway %s", networkID)
+	}
+
+	subProtosById := map[string]proto.Message{}
+	for _, sub := range subEnts {
+		subProto, err := convertSubEntsToProtos(sub, apnsByName, apnResourcesByAPN)
+		if err != nil {
+			return "", err
+		}
+		subProto.NetworkId = &protos.NetworkID{Id: networkID}
+
+		index := subProto.Sid.Id
+		subProtosById[index] = subProto
+	}
+
+	digest, err := mproto.MarshalManyDeterministic(subProtosById)
+	if err != nil {
+		return "", err
+	}
+	digestB64 := b64.StdEncoding.EncodeToString(digest)
+
+	return digestB64, nil
```

Comment:
Ah, something I realized we forgot to do in the previous PRs -- we want this library code to return a *digest*, so the hash of this encoded value. Cause right now the encoded value is O(n) (for n := num subscribers), while the digest would be O(1) (constant length)

Here's the example -- converting O(n) bytes to O(1) bytes using md5.Sum https://github.com/magma/magma/blob/509e5b4bcdb68b0021044fb1dc8cf47554446b8f/orc8r/lib/go/protos/mconfig_digest.go#L24-L35